File: GGAndroid_2.0/GGCore/GGVMCore/src/main/java/io/ganguo/vmodel/ViewModelActivity.java
package io.ganguo.vmodel;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;

import io.ganguo.library.ui.activity.BaseFragmentActivity;
import io.ganguo.library.ui.base.FragmentNavigator;
import io.ganguo.rx.ActivityResult;
import io.ganguo.rx.ActivityResultOwner;
import io.ganguo.utils.util.log.Logger;
import io.ganguo.library.ui.view.ActivityInterface;
import io.ganguo.library.ui.view.FragmentActivityInterface;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

/**
 * Base ViewModelActivity
 * Created by Roger on 6/7/16.
 *
 * @param <T> Activity ViewLayout Binding
 * @param <B> The ViewModel control the Activity
 */
public abstract class ViewModelActivity<T extends ViewDataBinding, B extends BaseViewModel>
        extends BaseFragmentActivity
        implements FragmentActivityInterface<T>, BaseViewModel.OnViewAttachListener<B>, ActivityResultOwner {
    private B mViewModel;
    private T mBinding;
    private Subject<ActivityResult> resultEmitter;

    public boolean checkViewModel() {
        boolean isNotNull = getViewModel() != null;
        if (!isNotNull) {
            Logger.e("viewModel is null");
        }
        return isNotNull;
    }

    @Override
    public void beforeInitView() {
        mBinding = DataBindingUtil.setContentView(this, getViewModel().getItemLayoutId());
    }

    @Override
    public void initView() {

    }

    @Override
    public void initListener() {
        getViewModel().setOnViewAttachListener(this);
    }

    @Override
    public void initData() {
        ViewModelHelper.bind((ActivityInterface) this, getViewModel());
    }

    @Override
    public Subject<ActivityResult> getActivityResult() {
        resultEmitter = PublishSubject.<ActivityResult>create().toSerialized();
        resultEmitter.hide();
        return resultEmitter;
    }

    /**
     * get ViewModel here
     *
     * @see #createViewModel()
     */
    public B getViewModel() {
        if (mViewModel == null) {
            mViewModel = createViewModel();
        }
        return mViewModel;
    }

    @Override
    public T getBinding() {
        return mBinding;
    }

    @Override
    public Context getContext() {
        return this;
    }

    @Override
    public Activity getActivity() {
        return this;
    }

    @Override
    public Bundle getBundle() {
        if (getIntent() == null) {
            return null;
        }
        return getIntent().getExtras();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (checkViewModel()) {
            getViewModel().getLifecycleHelper().onResume();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (checkViewModel()) {
            getViewModel().getLifecycleHelper().onPause();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (checkViewModel()) {
            getViewModel().getLifecycleHelper().onStop();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        resultEmitter = null;
        if (checkViewModel()) {
            getViewModel().getLifecycleHelper().onDestroy();
        }
    }

    @Override
    public FragmentNavigator getNavigator() {
        return super.getNavigator();
    }

    /**
     * Forward activity Result to viewModel
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultEmitter != null) {
            resultEmitter.onNext(new ActivityResult(requestCode, resultCode, data));
            resultEmitter.onComplete();
        }
    }

    /**
     * Create ViewModel here
     * if you want to getViewModel instance use {@link #getViewModel()} instead this
     */
    public abstract B createViewModel();

}

File: GGAndroid_3.0/core/frame/mvvm-v1/src/main/java/io/ganguo/viewmodel/core/view/ViewModelActivity.kt
package io.ganguo.viewmodel.core.view

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.FragmentActivity
import io.ganguo.core.ui.activity.BaseActivity
import io.ganguo.rx.ActivityResult
import io.ganguo.rx.ActivityResultOwner
import io.ganguo.viewmodel.core.BaseViewModel
import io.ganguo.viewmodel.core.BaseViewModel.OnViewAttachListener
import io.ganguo.viewmodel.core.ViewModelHelper
import io.ganguo.viewmodel.core.viewinterface.ActivityInterface
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject

/**
 * <pre>
 *     @author : leo
 *     time   : 2020/05/29
 *     desc   : Base ViewModelActivity
 * </pre>
 * @param T : ViewDataBinding
 * @param B : BaseViewModel<*>
 * @property binding T
 * @property context Context
 * @property activity FragmentActivity
 * @property bundle Bundle?
 * @property viewModel B
 * @property resultEmitter Subject<ActivityResult>?
 */
abstract class ViewModelActivity<T : ViewDataBinding, B : BaseViewModel<*>> :
        BaseActivity(), ActivityInterface<T>, OnViewAttachListener<B>, ActivityResultOwner {
    private var resultEmitter: Subject<ActivityResult>? = null
    override val binding: T by lazy {
        DataBindingUtil.setContentView(this, viewModel.layoutId) as T
    }
    override val context: Context by lazy {
        this
    }
    override val activity: FragmentActivity by lazy {
        this
    }
    val viewModel: B by lazy {
        createViewModel()
    }
    override val bundle: Bundle? by lazy {
        if (intent == null) {
            null
        } else {
            intent.extras
        }
    }

    init {
        bindLifecycle(viewModel.lifecycleHelper)
    }

    override fun beforeInitView() {
    }

    override fun initView() {}
    override fun initListener() {
        viewModel.onViewAttachListener = this as OnViewAttachListener<BaseViewModel<*>>
    }

    override fun initData() {
        ViewModelHelper.bind(this, viewModel)
    }

    override fun getActivityResult(): Subject<ActivityResult> {
        resultEmitter = PublishSubject.create<ActivityResult>().toSerialized()
        resultEmitter!!.hide()
        return resultEmitter!!
    }


    /**
     * Forward activity Result to viewModel
     * @param requestCode Int
     * @param resultCode Int
     * @param data Intent?
     */
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultEmitter != null) {
            resultEmitter!!.onNext(ActivityResult(requestCode, resultCode, data))
            resultEmitter!!.onComplete()
        }
    }

    /**
     * Create ViewModel here
     * if you want to getViewModel instance use [.getViewModel] instead this
     */
    abstract fun createViewModel(): B

    override fun onDestroy() {
        super.onDestroy()
        resultEmitter = null
    }
}